### ISSUE TITLE
Change Request: require("v8-compile-cache") conditionally

### ISSUE THREAD

**Issue by bartlomieju** [eslint/eslint#15330]
### ESLint version

8.2.0

### What problem do you want to solve?

As of v1.15 Deno supports Node.js compatibility code. ESLint can run in this compatibility mode without a problem. 

Here's a PR that adds example of running ESLint with Deno:
https://github.com/denoland/manual/pull/164

However Deno doesn't have support for V8 compile cache. Due to top-level `require("v8-compile-cache")` in `bin/eslint.js` there are some environmental variables being accessed to check if V8 compile cache should be disable. It means that users must run with `--allow-env` permissions which seems undesirable.

### What do you think is the correct solution?

Add a conditional for requiring `v8-compile-cache`.

Replacing
```js
// to use V8's code cache to speed up instantiation time
require("v8-compile-cache");
```
with
```js
// to use V8's code cache to speed up instantiation time
if (globalThis.Deno == undefined) {
  require("v8-compile-cache");
}
```
should be sufficient.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
If that’s all it takes to support Deno, I think we can do that.

My only question is about the correct check for the condition. Rather than saying “if this isn’t Deno, require the package” can we instead say “if it is Node, require the package”? Is there something present in Node that isn’t in Deno we can test? Maybe “process”?

**Comment by bartlomieju** [comment]
> If that’s all it takes to support Deno, I think we can do that.
> 
> My only question is about the correct check for the condition. Rather than saying “if this isn’t Deno, require the package” can we instead say “if it is Node, require the package”? Is there something present in Node that isn’t in Deno we can test? Maybe “process”?

I'm not sure, it seems to me that won't work - in "compat" mode Deno tries to emulate Node.js complately, which means globals like `process` are available. I couldn't find information about it, but does `eslint` distribute a browser version? If that's the case maybe we could check if we're in a browser environment (which Deno would mimic too)?

**Comment by nzakas** [comment]
This code is in the CLI, so it doesn't run in the browser. We just need to worry about where the CLI is run.

Ah that's interesting. So in compat mode Deno tries to look as much like Node.js as possible. Hmmmm.

Well, I'm not opposed to adding an explicit check for Deno to get this moving. Is there some way to run ESLint on Deno during CI so we can verify that it's working?

**Comment by bartlomieju** [comment]
@nzakas certainly! We just need to install Deno in CI (https://github.com/denoland/setup-deno)
and then follow example here: https://deno.land/manual@main/npm_nodejs/compatibility_mode#example

**Comment by nzakas** [comment]
Cool! If you're willing to put together a PR with the change and the CI, we can take a look. (note: soonest it would be released would be in two weeks.)

**Comment by bartlomieju** [comment]
> Cool! If you're willing to put together a PR with the change and the CI, we can take a look. (note: soonest it would be released would be in two weeks.)

Thanks! I'll open a PR over the weekend

**Comment by bartlomieju** [comment]
@nzakas I started working on the PR, one question though: do you want me to duplicate `test_on_node` CI step that runs on Deno, as a simple test like presented in the above Manual link would suffice? We're still working on getting `mocha` to work in Deno's compatibility mode so I'm not confident I would be able to run the whole CI under Deno (and I guess it might add too much time to your CI setup).

**Comment by nzakas** [comment]
I think we can just do a simple smoke test without running the whole CI suite. Maybe just run ESLint on itself?

```
deno run \
   --compat --unstable \
   --allow-read --allow-write=./ --allow-env \
   ./bin/eslint.js lib/ tests/
```

**Comment by bartlomieju** [comment]
I'm going to close this issue, when setting up CI in this (#15336) I discovered that some packages (eg. `ignore`) access `process.env` in other situations and effectively full access to environmental variables is required anyway. 

So it's not justified to add a single conditional, sorry for the buzz. Anyway, I'm happy that `eslint` can be used with Deno nowadays.
